sql/2021/security/feature_adoption_by_topN_technologies.sql

```diff
@@ -0,0 +1,93 @@
+#standardSQL
+# Determines to what extent the top-N technology drivers are responsible for the global adoption of different security features
+-- from https://stackoverflow.com/a/54835472
+CREATE TEMP FUNCTION array_slice(arr ARRAY<STRING>, start INT64, finish INT64)
+RETURNS ARRAY<STRING> AS (
+  ARRAY(
+    SELECT part FROM UNNEST(arr) part WITH OFFSET AS index
+    WHERE index BETWEEN start AND finish ORDER BY index
+  )
+);
+
+WITH app_headers AS (
+  SELECT
+    t._TABLE_SUFFIX AS client,
+    headername,
+    category,
+    app,
+    respOtherHeaders,
+    t.url AS url
+  FROM
+    `httparchive.summary_requests.2021_07_01_*` AS r
+  INNER JOIN
+    `httparchive.technologies.2021_07_01_*` AS t
+  ON
+    r._TABLE_SUFFIX = t._TABLE_SUFFIX AND
+    r.urlShort = t.url,
+    UNNEST(['Content-Security-Policy', 'Content-Security-Policy-Report-Only', 'Cross-Origin-Embedder-Policy', 'Cross-Origin-Opener-Policy', 'Cross-Origin-Resource-Policy', 'Expect-CT', 'Feature-Policy', 'Permissions-Policy', 'Referrer-Policy', 'Report-To', 'Strict-Transport-Security', 'X-Content-Type-Options', 'X-Frame-Options', 'X-XSS-Protection']) AS headername
+  WHERE
+    firstHtml AND
+    category IN UNNEST(['Blogs', 'CDN', 'Web frameworks', 'Programming languages', 'CMS', 'Ecommerce', 'PaaS', 'Security'])
+)
+
+SELECT
+  client,
+  headername,
+  topN,
+  ARRAY_TO_STRING(array_slice(top_apps, 0, topN - 1), ', ', 'NULL') AS topN_apps,
+  COUNT(DISTINCT IF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername, ' ')) AND CONCAT(category, '_', app) IN UNNEST(array_slice(top_apps, 0, topN - 1)), url, NULL)) AS freq_in_topN,
+  SAFE_DIVIDE(COUNT(DISTINCT IF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername, ' ')) AND CONCAT(category, '_', app) IN UNNEST(array_slice(top_apps, 0, topN - 1)), url, NULL)), global_freq) AS pct_overall
```

Comment:
Some funky stuff going on here! Not got time to delve in now, so will need to come back to this to make sure it's right...

Reply:
Going to leave this for now as same as last year, but think it could be made clearer what's going on here. But as we used it last year let's just go with it for now.